File: src/test/java/info/kfgodel/dyna/dyna/AccessOwnStateTest.java
package info.kfgodel.dyna.dyna;

import ar.com.dgarcia.javaspec.api.JavaSpec;
import ar.com.dgarcia.javaspec.api.JavaSpecRunner;
import info.kfgodel.dyna.WorldTestContext;
import info.kfgodel.dyna.impl.DefaultEnvironment;
import org.junit.runner.RunWith;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Date: 05/05/19 - 17:44
 */
@RunWith(JavaSpecRunner.class)
public class AccessOwnStateTest extends JavaSpec<WorldTestContext> {
  @Override
  public void define() {
    describe("a dyna object", () -> {
      test().objectWithState(() -> DefaultEnvironment.create().creator().create(SelfStateObject.class));

      it("can introspect its own state map", () -> {
        assertThat(test().objectWithState().getPropertyCount()).isEqualTo(1);

        test().objectWithState().setName("Pepe");
        assertThat(test().objectWithState().getPropertyCount()).isEqualTo(2);
      });


      it("can change a method definition",()->{
        assertThat(test().objectWithState().getName()).isNull();

        test().objectWithState().changeGetNameDefinition();

        assertThat(test().objectWithState().getName()).isEqualTo("a hardcoded name");
      });
    });

  }
}

File: src/test/kotlin/info/kfgodel/dyna/state/AccessOwnStateTest.kt
package info.kfgodel.dyna.state

import info.kfgodel.dyna.dyna.SelfStateObject
import info.kfgodel.dyna.impl.DefaultEnvironment
import info.kfgodel.jspek.api.JavaSpecRunner
import info.kfgodel.jspek.api.KotlinSpec
import org.assertj.core.api.Assertions
import org.junit.runner.RunWith

/**
 * Date: 28/3/21 - 02:00
 */
@RunWith(JavaSpecRunner::class)
class AccessOwnStateTest : KotlinSpec() {
  override fun define() {
    describe("a dyna object") {
      val objectWithState by let {DefaultEnvironment.create().creator().create(
        SelfStateObject::class.java
      )}

      it("can introspect its own state map") {
        Assertions.assertThat(objectWithState().getPropertyCount()).isEqualTo(1)
        objectWithState().name = "Pepe"
        Assertions.assertThat(objectWithState().getPropertyCount()).isEqualTo(2)
      }
      it("can change a method definition") {
        Assertions.assertThat(objectWithState().name).isNull()
        objectWithState().changeGetNameDefinition()
        Assertions.assertThat(objectWithState().name).isEqualTo("a hardcoded name")
      }
    }

  }
}
